common/transport/amqp/src/amqp.ts

```diff
@@ -205,38 +306,39 @@ export class Amqp {
    * @param {Function}   done   Called when disconnected of if an error happened.
    */
   disconnect(done: GenericAmqpBaseCallback): void {
-    const self = this;
+    this._fsm.handle('disconnect', done);
+  }
 
+  _disconnect(done: GenericAmqpBaseCallback): void {
     /*Codes_SRS_NODE_COMMON_AMQP_16_034: [The `disconnect` method shall detach all open links before disconnecting the underlying AMQP client.]*/
-    let openLinks = [];
     for (let senderEndpoint in this._senders) {
       if (this._senders.hasOwnProperty(senderEndpoint)) {
-        openLinks.push(this._senders[senderEndpoint]);
+        this._senders[senderEndpoint].detach();
         delete this._senders[senderEndpoint];
       }
     }
-    self._senders = {};
+    this._senders = {};
 
     for (let receiverEndpoint in this._receivers) {
       if (this._receivers.hasOwnProperty(receiverEndpoint)) {
-        openLinks.push(this._receivers[receiverEndpoint]._amqpReceiver);
+        this._receivers[receiverEndpoint].detach();
         delete this._receivers[receiverEndpoint];
       }
     }
-    self._receivers = {};
-
-    Promise.all(openLinks.map((link) => link.detach()))
-          .then(() => this._amqp.disconnect())
-          .then((result) => {
-            self._connected = false;
-            /*Codes_SRS_NODE_COMMON_AMQP_16_004: [The disconnect method shall call the done callback when the application/service has been successfully disconnected from the service] */
-            self._safeCallback(done, null, result);
-            return null;
-          })
-          .catch((err) => {
-            /*SRS_NODE_COMMON_AMQP_16_005: [The disconnect method shall call the done callback and pass the error as a parameter if the disconnection is unsuccessful] */
-            self._safeCallback(done, err);
-          });
+    this._receivers = {};
+
+    if (this._cbs) {
+      this._cbs.detach();
+    }
```

Comment:
So, was it really best that the above detaches in the for loop killed the senders and receivers out from under the cbs before the cbs detach code did it?